third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+) -> Option<Vec<u8>> {
+    let native_serializer = NativeValueSimpleSerDe;
+    let value = SerializationReadyValue {
+        native_serializer: Some(&native_serializer),
+        layout,
+        value: &value.0,
+    };
+    bcs::to_bytes(&value).ok()
+}
+
+pub trait ValueToIdentifierMapping {
+    fn value_to_identifier(
+        &self,
+        // We need kind to distinguish between aggregators and snapshots
+        // of the same type.
+        kind: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        value: Value,
+    ) -> PartialVMResult<SizedID>;
+
+    fn identifier_to_value(
+        &self,
+        layout: &MoveTypeLayout,
+        identifier: SizedID,
+    ) -> PartialVMResult<Value>;
+}
+
+pub struct NativeValueSerDeWithExchange<'a> {
+    mapping: &'a dyn ValueToIdentifierMapping,
+}
+
+impl<'a> CustomSerialize for NativeValueSerDeWithExchange<'a> {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        _tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error> {
+        let value = self
+            .mapping
+            .identifier_to_value(layout, sized_id)
+            .map_err(|e| S::Error::custom(format!("{}", e)))?;
+        SerializationReadyValue {
+            native_serializer: None::<&NativeValueSimpleSerDe>,
+            layout,
+            value: &value.0,
+        }
+        .serialize(serializer)
+    }
+}
+
+impl<'a> CustomDeserialize for NativeValueSerDeWithExchange<'a> {
+    fn custom_deserialize<'d, D: Deserializer<'d>>(
+        &self,
+        deserializer: D,
+        tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+    ) -> Result<Value, D::Error> {
+        let value = DeserializationSeed {
+            native_deserializer: None::<&NativeValueSimpleSerDe>,
+            layout,
+        }
+        .deserialize(deserializer)?;
+        let size_id = self
+            .mapping
+            .value_to_identifier(tag, layout, value)
+            .map_err(|e| D::Error::custom(format!("{}", e)))?;
+        Ok(Value::native_value(size_id))
+    }
+}
+
+pub fn deserialize_and_replace_values_with_ids(
+    bytes: &[u8],
+    layout: &MoveTypeLayout,
+    mapping: &impl ValueToIdentifierMapping,
+) -> Option<Value> {
+    let native_deserializer = NativeValueSerDeWithExchange { mapping };
+    let seed = DeserializationSeed {
+        native_deserializer: Some(&native_deserializer),
+        layout,
+    };
+    bcs::from_bytes_seed(seed, bytes).ok()
+}
+
+pub fn serialize_and_replace_ids_with_values(
+    value: &Value,
+    layout: &MoveTypeLayout,
+    mapping: &impl ValueToIdentifierMapping,
+) -> Option<Vec<u8>> {
+    let native_serializer = NativeValueSerDeWithExchange { mapping };
+    let value = SerializationReadyValue {
+        native_serializer: Some(&native_serializer),
+        layout,
+        value: &value.0,
+    };
+    bcs::to_bytes(&value).ok()
```

Comment:
So, we are basically storing the `native_serializer` code and layout along with the value when serializing. When deserializing, we have to remove these and extract only the value. Where are we doing that?

Reply:
Yes! or to be more precise, references to them. Then serialization can look at these values and decide on how to serialize things and when to error.

For deserialization, we use a different struct (`seed`) with the same purpose, it allows us to have information on the side about how to interpret the bytes.